packages/dd-trace/src/appsec/downstream_requests.js

```diff
@@ -0,0 +1,323 @@
+ */
+function extractResponseData (res, responseBody) {
+  const addresses = {}
+
+  if (res.statusCode !== undefined && res.statusCode !== null) {
+    addresses[HTTP_OUTGOING_RESPONSE_STATUS] = String(res.statusCode)
+  }
+
+  const headers = res.headers
+  if (headers && Object.keys(headers).length > 0) {
+    addresses[HTTP_OUTGOING_RESPONSE_HEADERS] = headers
+  }
+
+  if (responseBody) {
+    // Parse the body based on content-type
+    const contentType = getResponseContentType(res.headers)
+    const body = parseBody(responseBody, contentType)
+
+    if (body !== null && body !== undefined) {
+      addresses[HTTP_OUTGOING_RESPONSE_BODY] = body
+    }
+  }
+
+  return addresses
+}
+
+/**
+ * Tracks how many downstream analyses were executed for a given request and updates tracing tags.
+ * @param {import('http').IncomingMessage} req inbound request.
+ */
+function incrementDownstreamAnalysisCount (req) {
+  const currentCount = downstreamAnalysisCount.get(req) || 0
+  downstreamAnalysisCount.set(req, currentCount + 1)
+
+  const span = web.root(req)
+
+  if (span) {
+    span.setTag('_dd.appsec.downstream_request', currentCount + 1)
+  }
+}
+
+/**
+ * Adds tracing telemetry for matches triggered by downstream responses.
+ * @param {import('http').IncomingMessage} req inbound reques.
+ * @param {object} raspRule rule with type and variant
+ */
+function handleResponseTracing (req, raspRule) {
+  updateRaspRuleMatchMetricTags(req, raspRule, false, false)
+}
+
+/**
+ * Returns the HTTP method to use for a downstream request, defaulting to GET.
+ * @param {string} method method supplied in the outgoing request options.
+ * @returns {string} validated HTTP method.
+ */
+function determineMethod (method) {
+  return typeof method === 'string' && method ? method : 'GET'
+}
+
+/**
+ * Gets the content-type header value from a case-insensitive headers object.
+ * @param {import('http').IncomingHttpHeaders|object|null} headers response headers object.
+ * @returns {string|null} content-type value
+ */
+function getResponseContentType (headers) {
+  if (!headers) return null
+
+  return headers['content-type'] || headers['Content-Type'] || headers['CONTENT-TYPE'] || null
+}
+
+/**
+ * Parses a downstream response body.
+ * @param {Buffer|string|object|null} body raw response body
+ * @param {string|null} contentType response content-type used to select the parser.
+ * @returns {object|null} parsed body object or null when not supported.
+ */
+function parseBody (body, contentType) {
+  if (body === null || body === undefined || !contentType) {
+    return null
+  }
+
+  const mime = extractMimeType(contentType)
+
+  try {
+    if (mime === 'application/json' || mime === 'text/json') {
+      if (typeof body === 'string') {
+        return JSON.parse(body)
+      }
+
+      if (Buffer.isBuffer(body)) {
+        return JSON.parse(body.toString('utf8'))
+      }
+
+      if (typeof body === 'object' && body !== null) {
```

Comment:
can this really be an object anyway ? maybe it can, but in that case the urlencoded if could also be an object no ? either both can be or both cannot be ?